ipc/provider/src/manager/evm/gas_estimator_middleware.rs

```diff
@@ -0,0 +1,160 @@
+        let max_priority_fee_per_gas = self
+            .max_priority_fee_per_gas()
+            .await
+            .map_err(|e| Eip1559GasEstimatorError::FailedToEstimateGas(e.to_string()))?;
+        let base_fee_per_gas = self
+            .base_fee_per_gas()
+            .await
+            .map_err(|e| Eip1559GasEstimatorError::FailedToEstimateGas(e.to_string()))?;
+
+        // Buffer the base fee by multiplying by 2 to account for potential cumulative increases.
+        let base_fee_per_gas_surged = base_fee_per_gas * 2;
+        let max_fee_per_gas = max_priority_fee_per_gas + base_fee_per_gas_surged;
+
+        Ok(GasFeeEstimate {
+            max_priority_fee_per_gas,
+            max_fee_per_gas,
+        })
+    }
+}
+
+#[async_trait]
+impl<M: Middleware> Middleware for Eip1559GasEstimatorMiddleware<M> {
+    type Error = Eip1559GasEstimatorError<M>;
+    type Provider = M::Provider;
+    type Inner = M;
+
+    fn inner(&self) -> &M {
+        &self.inner
+    }
+
+    /// Fills the transaction with EIP-1559 gas fees.
+    async fn fill_transaction(
+        &self,
+        tx: &mut TypedTransaction,
+        block: Option<BlockId>,
+    ) -> Result<(), Self::Error> {
+        if let TypedTransaction::Eip1559(inner) = tx {
+            let gas_fees = self.estimate_gas_fees().await?;
+
+            // Set the gas fees directly on the transaction.
+            let tx_req = inner
+                .clone()
+                .max_fee_per_gas(gas_fees.max_fee_per_gas)
+                .max_priority_fee_per_gas(gas_fees.max_priority_fee_per_gas);
+
+            *tx = TypedTransaction::Eip1559(tx_req);
+        } else {
+            return Err(Eip1559GasEstimatorError::failed_to_estimate_gas_not_supported());
+        }
+
+        // Delegate to the inner middleware for filling remaining transaction fields.
+        self.inner()
+            .fill_transaction(tx, block)
+            .await
+            .map_err(Eip1559GasEstimatorError::MiddlewareError)
+    }
+
+    /// Sends a transaction with EIP-1559 gas fees.
+    async fn send_transaction<T: Into<TypedTransaction> + Send + Sync>(
+        &self,
+        tx: T,
+        block: Option<BlockId>,
+    ) -> Result<PendingTransaction<'_, Self::Provider>, Self::Error> {
+        let mut tx = tx.into();
+
+        // Automatically fill EIP-1559 gas fees if they are not already set.
+        if let TypedTransaction::Eip1559(ref mut inner) = tx {
+            if inner.max_fee_per_gas.is_none() || inner.max_priority_fee_per_gas.is_none() {
+                // Populate missing gas fees with `fill_transaction`.
+                self.fill_transaction(&mut tx, block).await?;
+            }
+        } else {
+            return Err(Eip1559GasEstimatorError::failed_to_estimate_gas_not_supported());
+        }
+
+        // Proceed to send the transaction with the inner middleware.
+        self.inner()
+            .send_transaction(tx, block)
```

Comment:
same question here, the inner will not overwrite the gas params updated right?

Reply:
Nope it should not. Answer above.👆🏻